Implemented CPU plugin just-in-time emitter for SoftPlus operation.

geeky33 opened the pull request:
### Details:
 # Add ARM64 JIT emitter implementation for SoftPlus operation (ln(1 + exp(x))) with fp32 support.
 - Changes include:
-  *Implement SoftPlus JIT emitter using ARM NEON instructions*
-  *Add kernel support in ARM64 executor*
- *Add test coverage for JIT SoftPlus operation*
- *Included comprehensive unit tests and performance benchmarks. Verified against reference CPU implementation with proper handling of edge cases (NaN, infinity, denormals).*

No breaking changes. Requires ARM64 platform with NEON SIMD support.

 @eshoguli 
 @dmitry-gorokhov 
 @a-sidorova 
 @p-wysocki 

### Tickets:
 -  *#24109*

a-sidorova commented on src/plugins/intel_cpu/src/emitters/plugin/aarch64/jit_eltwise_emitters.cpp:
```diff
@@ -2501,6 +2501,67 @@ void jit_select_emitter::emit_isa(const std::vector<size_t>& in_vec_idxs,
+    const TReg aux1 = TReg(aux_vec_idxs[0]);
+    const TReg aux2 = TReg(aux_vec_idxs[1]);
+
+    h->fexp(aux1.s, src.s);
+    h->ld1r(aux2.s, table_val2("one"));
+    h->fadd(aux1.s, aux1.s, aux2.s);
+    h->flog(dst.s, aux1.s);
```
As far as I remember, you told that you have problems with test launching using qemu (but as we agreed, I don't see comments with asking help with your problem in your GFI).
But anyway do you have a chance to validate this implementation? I didn't find the instructions in `fexp` or `flog` in [AArch64 manual](https://developer.arm.com/documentation/ddi0602/2024-03/SIMD-FP-Instructions?lang=en).

Exp can be calculated used `jit_exp_emitter` (please see the [`Sigmoid`](https://docs.openvino.ai/2024/documentation/openvino-ir-format/operation-sets/operation-specs/activation/sigmoid-1.html) emitter implementation, this emitter [needs to calculate exp](https://github.com/openvinotoolkit/openvino/blob/fb1838fbe8d2d1054ff331c5958397b30df5df6d/src/plugins/intel_cpu/src/emitters/plugin/aarch64/jit_eltwise_emitters.cpp#L2506) too).

As for `log`, it can be calculated using approximation. Please take a look at [oneDNN impl](https://github.com/oneapi-src/oneDNN/blob/main/src/cpu/aarch64/injectors/jit_uni_eltwise_injector.cpp#L915). But it's implemented using SVE. You can try to rewrite it using NEON with the same math underneath

geeky33 replied:
Hello Ma'am
I realised that this was the PR which I created after talking with you realted to launching the tests issue. Hence I thought to refactor the code after getting your review and then ask in the comments for asking help on the tests.

a-sidorova commented on src/plugins/intel_cpu/src/nodes/executors/aarch64/jit_eltwise.cpp:
```diff
@@ -57,6 +57,7 @@ bool JitEltwiseExecutor::isSupported(const Algorithm& algorithm,
                                      Algorithm::EltwiseRoundHalfAwayFromZero,
                                      Algorithm::EltwiseRoundHalfToEven,
                                      Algorithm::EltwiseSelect,
+                                     Algorithm::EltwiseSoftPlus,
```
Looks like `EltwiseSoftPlus` is missed in `Algorithm` enum

geeky33 commented:
@a-sidorova 
Greetings Ma'am

I have done the changes according to the comments made.

Please let me know what all more changes are needed to be done further.

Thanks and Regards.

a-sidorova commented:
@geeky33 if contributor asks to review, I assume that the contributor is sure that code is correct and works.  And I may spend my time on review. Otherwise, contributor should ask help with the code. I already left the [comment](https://github.com/openvinotoolkit/openvino/pull/28752#discussion_r1936676809) with helpful links and hints how to implement `SoftPlus` emitter with `exp` and `log`. Unfortunately, I don't see applied changes (removing `fexp` and replacing `flog` with `flog_approx` don't fix the implementation) but you asked me to review it again :)
If you have concrete questions about the implementation and my previous comment doesn't help, please let me know and feel free to ask!

As for testing and emulator problems, I asked you to leave comment in the [issue](https://github.com/openvinotoolkit/openvino/issues/24109) with asking help with faced problem. I still don't see this comment with problem description and your steps for reproduction.

geeky33 commented:
Greetings once again.
I will be making the exact changes by next few hours ma'am
Extremely sorry for the issue being caused
I am on it.

Now, since launching the tests was an issue to me

Here are the steps I followed for NotEqual JIT Emitter


```
sudo apt-get update
sudo apt-get install qemu-user qemu-user-static gcc-aarch64-linux-gnu g++-aarch64-linux-gnu binutils-aarch64-linux-gnu
set(CMAKE_SYSTEM_NAME Linux)
set(CMAKE_SYSTEM_PROCESSOR aarch64)

set(CMAKE_C_COMPILER aarch64-linux-gnu-gcc)
set(CMAKE_CXX_COMPILER aarch64-linux-gnu-g++)
set(CMAKE_STRIP aarch64-linux-gnu-strip)
set(PKG_CONFIG_EXECUTABLE aarch64-linux-gnu-pkg-config CACHE PATH "Path to ARM64 pkg-config")
sudo apt-get install -y \
    qemu-user \
    qemu-user-static \
    gcc-aarch64-linux-gnu \
    g++-aarch64-linux-gnu \
    binutils-aarch64-linux-gnu \
    build-essential \
    cmake
aarch64-linux-gnu-gcc --version
aarch64-linux-gnu-g++ --version

sudo apt update
sudo apt install gcc-aarch64-linux-gnu g++-aarch64-linux-gnu binutils-aarch64-linux-gnu libc6-dev-arm64-cross
```

Till here everythhing went superbly fine 

I have ran the build cmd again after deleting the dir
and i m facing the same issue again 

![image](https://github.com/user-attachments/assets/4d9dd2f4-c9fe-487e-a153-493ffc8c83d5)

![image](https://github.com/user-attachments/assets/1446d5b5-7a48-4aa8-9cb8-148b28645001)


ran this cmd `sudo apt install libc6:aarch64`

and also followed your instructions 
removed build and bin directories
updated aarch64 compiler
And tried to configure openvino and built it one more time

even after subsequent tries I can see that the build is completed with 100% even after that I am not able to launch the relevant testcases.

geeky33 commented:
Extremely sorry for the issue being caused:
 
  Before making any changes,
  could you please just check this part of the codeblock which I am attaching below:
  
  
  ```
    using TReg = typename dnnl::impl::cpu::aarch64::cpu_isa_traits<isa>::TReg;
    const TReg src(in_vec_idxs[0]);
    const TReg dst(out_vec_idxs[0]);
    const TReg aux1(aux_vec_idxs[0]);
    const TReg aux2(aux_vec_idxs[1]);
    const TReg aux3(aux_vec_idxs[2]);
    const TReg aux4(aux_vec_idxs[3]);
    const TReg aux5(aux_vec_idxs[4]);

    h->ld1r(aux1.s, table_val2("threshold"));
    h->ld1r(aux3.s, table_val2("one"));
    h->fcmgt(aux2.s, src.s, aux1.s);
    h->fmov(aux4.s, src.s);
    h->fadd(aux4.s, aux4.s, aux3.s);
    h->fmov(aux5.s, aux4.s);
    h->bsl(dst.s, aux2.s, src.s, aux5.s);
```


To validate my this implementation:
This implementation appears to be appropriate for performing a SIMD thresholding operation: it compares `src` with a threshold, increments values below the threshold by 1.0, and uses `bsl` to conditionally select between the original and incremented values. Ensuring that table_val2 correctly loads `threshold` and `one`, and validate edge cases (e.g., threshold equality, NaN).

a-sidorova commented:
@geeky33  First of all, I suggest to fix your problem with qemu. As I can see in your previous screenshot, you have the same problem when you try to execute even simple `void main() {}` using qemu. So I think there is some problem in your environment, not in OpenVINO. Launching of tests is the one of important parts in development :)

Then let's do so: when you are sure in your implementation, launch all needed tests for `SoftPlus` and they are green in your local environment, please attach screenshots that all tests are successfully passed on your machine and ask me to review your PullRequest. When I see that everything is ready and tests are successfully and locally passed on your machine (see your attached screenshot), I will take a look at your changes.

Also, I asked my colleagues if someone of them faced with the same problem in launching tests with qemu, but no one see it in his environment. So I think that firstly you need to spend some time on googling if you want to launch tests with emulator :)

geeky33 commented:
@a-sidorova Thank you for the feedback
Is it possible to get in touch with any of your colleague who looks after the setup areas of openvino for fixing this issue?
Since it is very overwhelming for me to fix this as I tried doing research and I could not find enough clues.
I believe that I have tried all the possible attempts to fix this issue from my side and also the suggestions given by you in the discord

Still whatever could be the done I will look into it as well!  :)

a-sidorova commented:
Closed the PR since now we have another assignee of this task.

@geeky33 respect for the attempt! ❤️

geeky33 commented:
@a-sidorova thanks for being patient with me throughout ma'am❤️🫶
